### PR TITLE
[FLINK-16336][table] Add new type inference for temporal table functions

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

Uses the new type system with new type inference in temporal table functions.

Depends on FLINK-16261.

## Brief change log

Function now implements both `getResultType()` and `getTypeInference()`.

## Verifying this change

This change is already covered by existing tests, such as `TemporalJoinITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala [hunk 1]
```diff
@@ -562,6 +562,13 @@ abstract class TableTestUtil(
     testingTableEnv.scan(name)
   }
 
+  /**
+   * Registers a [[UserDefinedFunction]] according to FLIP-65.
+   */
+  def addTemporarySystemFunction(name: String, function: UserDefinedFunction): Unit = {
```
Comment (aljoscha) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala]: Why is this added? It's not used anywhere?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/logical/FlinkLogicalTableFunctionScan.scala [hunk 2]
```diff
@@ -96,11 +96,12 @@ class FlinkLogicalTableFunctionScanConverter
       return false
     }
     val rexCall = logicalTableFunction.getCall.asInstanceOf[RexCall]
-    if (!rexCall.getOperator.isInstanceOf[TableSqlFunction]) {
-      return false
+    val functionDefinition = rexCall.getOperator match {
+      case tsf: TableSqlFunction => tsf.udtf
+      case bsf: BridgingSqlFunction => bsf.getDefinition
```
Comment (aljoscha) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/logical/FlinkLogicalTableFunctionScan.scala]: This is just for my understanding, but where is the bridging function created?

Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/logical/FlinkLogicalTableFunctionScan.scala]: In the catalog layer, see `FunctionCatalogOperatorTable`.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/TableTestBase.scala]: Good point, I will add usage. It is definitely necessary to have in the utils otherwise we have no means to test new functions.

**Comment by aljoscha** [comment]
Lools good to merge!

**Comment by twalthr** [comment]
@flinkbot run travis
